docs: Fr/sequential example (#156)
best-practice-and-impact/afcharts-py: src/afcharts/cookbook/03-plotly-usage.qmd

<div class="chart-source">Source: Gapminder</div>

This bar chart uses the afcharts theme, and shows the populations of five countries of the Americas in descending order. The country names are given on the x axis, with all chart text in black in a sans serif font. Four of the bars on the chart are light grey, and the bar for Brazil is filled in dark blue to highlight it.

## Sequential palette use example

```{python}
# | eval: true
# | output: false

import requests
import pandas as pd
import geopandas as gpd
import plotly.graph_objects as go
import io

# Fetch indicator data
url = "https://fingertips.phe.org.uk/api/all_data/csv/by_indicator_id?indicator_ids=41203"
df = pd.read_csv(io.BytesIO(requests.get(url, verify=False).content))

# Keep most recent period
df = df[df["Time period"] == "2023/24"]  # This is the date in the analysis function website.

> # df = df.sort_values("Time period").drop_duplicates("Area Code", keep="last")

Author: I left this here for us to discuss weather to use the latest available data or the same year as the AF website. @nickhenden

> # df = df.sort_values("Time period").drop_duplicates("Area Code", keep="last")

Reviewer: Let's use the same year as the AF website so it stays the same. We don't want our website to break because there's something dodgy with the latest data

bins = [0, 2.4, 2.9, 3.5, 4.6, 12.7]

# Labels in the same order
labels = ["Lowest (0.0 to 2.4)", "(2.4 to 2.9)", "(2.9 to 3.5)", "(3.5 to 4.6)", "Highest (4.6 to 12.7)"]
order = ["Suppressed"] + labels

# Assign bin categories
df["quantile_label"] = pd.cut(df["Value"], bins=bins, labels=labels, include_lowest=True)

# Handle suppressed values (e.g., NaN)
df["quantile_label"] = df["quantile_label"].astype("category")
df["quantile_label"] = df["quantile_label"].cat.add_categories("Suppressed")
df["quantile_label"] = df["quantile_label"].fillna("Suppressed")

df["quantile_label"] = df["quantile_label"].cat.reorder_categories(order, ordered=True)

# Get UK Country boundaries
url = (
    "https://services1.arcgis.com/ESMARspQHYMw9BZ9/arcgis/rest/services/"
    "Countries_December_2023_Boundaries_UK_BUC/FeatureServer/0/query"
    "?outFields=*&where=1%3D1&f=geojson"
)
country_boundaries = gpd.read_file(requests.get(url, verify=False).content).to_crs(4326)

# Get Counties and Unitary Autorities
url = "https://services1.arcgis.com/ESMARspQHYMw9BZ9/arcgis/rest/services/Counties_and_Unitary_Authorities_December_2023_Boundaries_UK_BUC/FeatureServer/0/query?outFields=*&where=1%3D1&f=geojson"

> gdf = gpd.read_file(requests.get(url, verify=False).content).to_crs(4326)

Author: @nickhenden I need you to test this your end. withouth it my enviroment can't make requests.

Oldham local authority has the highest rate where 12.7 per 100,000 persons aged 12 years and over were newly certified with a visual impairment due to diabetic eye disease. In contrast City of London local authority has the lowest rate of 0.0 per 100,000 persons aged 12 years and over. 29 local authority areas had values that were suppressed due to a small count.

## Heat Maps
```{python}
# | eval: true
# | output: false

import pandas as pd
import plotly.graph_objects as go

import textwrap

from afcharts.af_colours import get_af_colours
from afcharts.pio_template import pio

# Get the sequential colour palette
sequential = get_af_colours("sequential")[0:-1][::-1]

# Set default theme
pio.templates.default = "afcharts"

# Source: https://www.gov.uk/government/statistical-data-sets/monthly-domestic-energy-price-statistics

> df = pd.read_excel("./data/table_211_213__6_.xlsx", sheet_name="2.1.1", header=11)

Author: url was wrong but I got it to work

df = df.set_index(["Year and Period"])

# Drop row 0 (contains repeated header artefacts), select numeric block
df = df.iloc[1:, 9:-2].copy()

for c in df.columns:
    df[c] = pd.to_numeric(df[c], errors="coerce")

fig = go.Figure()

# Wrap each x‑axis label to max 30 characters per line
wrapped_x = ["<br>".join(textwrap.wrap(col.replace("Real terms price indices:", ""), width=30)) for col in df.columns]

fig.add_trace(go.Heatmap(z=df, x=wrapped_x, y=list(df.index), colorscale=sequential))

fig.update_layout(
    height=600,
    xaxis=dict(
        tickangle=45,
        automargin=True,
    ),
)

fig.show()
```

<div class="chart-title">Domestic energy price indices in real terms</div>
<div class="chart-subtitle">Heat map of Quarterly consumer price indices of fuel components.</div>

```{python}
#| echo: false

fig.show()

```

<div class="chart-source">Source: https://www.gov.uk/government/statistical-data-sets/monthly-domestic-energy-price-statistics</div>

Figures include VAT where applicable.  VAT rates for coal and coke, gas, electricity and heating oils was 8% from the 2nd quarter of 1994 and 5% since the 4th quarter of 1997 (the rate changed on 1st September). 

> Real price series are deflated using GDP (market prices) deflator. GDP updated is updated when published by ONS.

Author: ahahah I copied this straight from the file 😂 did not even read it.

<div class="chart-source">Source: Gapminder</div>

This bar chart uses the afcharts theme, and shows the populations of five countries of the Americas in descending order. The country names are given on the x axis, with all chart text in black in a sans serif font. Four of the bars on the chart are light grey, and the bar for Brazil is filled in dark blue to highlight it.

> ## Sequential palette use example

Reviewer: Suggest renaming so it's clear it's a map, e.g.

```suggestion
## Choropleth map with sequential palette
```

Oldham local authority has the highest rate where 12.7 per 100,000 persons aged 12 years and over were newly certified with a visual impairment due to diabetic eye disease. In contrast City of London local authority has the lowest rate of 0.0 per 100,000 persons aged 12 years and over. 29 local authority areas had values that were suppressed due to a small count.
</div>

## Heat map with sequential palette
```{python}
# | eval: true
# | output: false

import pandas as pd
import plotly.graph_objects as go

import textwrap

from afcharts.af_colours import get_af_colours
from afcharts.pio_template import pio

# Get the sequential colour palette
sequential = get_af_colours("sequential", number_of_colours=5)[::-1]

# Set default theme
pio.templates.default = "afcharts"

# Source: https://www.gov.uk/government/statistical-data-sets/monthly-domestic-energy-price-statistics

> DATA_URL = "https://assets.publishing.service.gov.uk/media/699c5d49cd3d70e6c2f1274c/table_211_213__6_.xlsx"

Reviewer: Annoyingly, the URL has changed. 
```suggestion
DATA_URL = "https://assets.publishing.service.gov.uk/media/69f1caf7c42061e837e3abfb/table_211_213__8_.xlsx"
```

> df = pd.read_excel(DATA_URL, sheet_name="2.1.1", header=11)

Reviewer: This excel file has an annoying second header row. To get this to work I had to add `skiprows=[12]` to skip that second header row.
```suggestion
df = pd.read_excel(DATA_URL, sheet_name="2.1.1", header=11, skiprows=[12])
```

gdf = gpd.read_file(requests.get(url, verify=False).content).to_crs(4326)

# Filter to England only
england_boundaries = country_boundaries[country_boundaries["CTRY23NM"] == "England"]
gdf_clipped_strict = gpd.overlay(gdf, england_boundaries, how="intersection")

merged = gdf_clipped_strict.merge(df, left_on="CTYUA23CD", right_on="Area Code", how="left").reset_index(drop=True)

from afcharts.af_colours import get_af_colours
from afcharts.pio_template import pio

# Get the sequential colour palette in reverse order
sequential = get_af_colours("sequential", number_of_colours=5, include_grey=True)[::-1]

# Set default theme
pio.templates.default = "afcharts"

fig = go.Figure()

cat_to_code = {label: i + 1 for i, label in enumerate(labels)}
z = merged["quantile_label"].map(cat_to_code)

# Must add a separate trace for the 'Suppressed'.
null_mask = ~pd.isna(z)
fig.add_trace(
    go.Choropleth(
        geojson=merged.loc[null_mask].__geo_interface__,
        locations=merged.loc[null_mask].index,
        z=z[null_mask],
        colorscale=sequential[1:],
        zmin=0,
        zmax=5,

>         marker_line_color="white",

Reviewer: Reduce line width to what it was before (0.5):

```suggestion
        marker_line_width=0.5,
        marker_line_color="white",
```

</div>

## Heatmap
<!-- Source: https://www.gov.uk/government/statistical-data-sets/monthly-domestic-energy-price-statistics -->
```{python}
# | eval: true
# | output: false

import numpy as np
import pandas as pd
import plotly.graph_objects as go

import textwrap

from afcharts.af_colours import get_af_colours
from afcharts.pio_template import pio

# Get the sequential colour palette
sequential = get_af_colours("sequential", number_of_colours=5, include_grey=True)[::-1]

# Set default theme
pio.templates.default = "afcharts"

DATA_URL = "https://assets.publishing.service.gov.uk/media/69f1caf7c42061e837e3abfb/table_211_213__8_.xlsx"

df = pd.read_excel(DATA_URL, sheet_name="2.1.1", header=11, skiprows=[12], usecols="A:B,J:P")

df = (
    df[df["Quarter"] == "Jan to Mar"]
    .set_index("Year and dataset code row")
    .drop(columns=["Quarter"])
    .dropna(how="all")
)
df.index = df.index.astype("string")

df.columns = (
    df.columns.str.replace("Real terms price indices: ", "")
    .str.replace(" [Note 3]", "")
    .str.replace("\n[Note 1, 3]", "")
    .str.replace("\n[Note 2, 3]", "")

>     .str.replace("=100\n[Note 3]", "")

Reviewer: Should just be "All items". The "CPI 2025=100" suffix is the units, which should be on the colourbar.
```suggestion
    .str.replace("CPI 2025=100\n[Note 3]", "")
```

)

for c in df.columns:
    df[c] = pd.to_numeric(df[c], errors="coerce")

fig = go.Figure()

fig.add_trace(
    go.Heatmap(
        z=df.T,
        x=df.index,
        y=df.columns,

>         colorscale=sequential[1:],

Reviewer: It wasn't obvious what the colours meant. Suggest you add a label to the colour bar showing its the Consumer Price Index (where 2025 is 100)

```suggestion
        colorscale=sequential[1:],
        colorbar=dict(title="CPI (2025 = 100)"),
```

This bar chart uses the afcharts theme, and shows the populations of five countries of the Americas in descending order. The country names are given on the x axis, with all chart text in black in a sans serif font. Four of the bars on the chart are light grey, and the bar for Brazil is filled in dark blue to highlight it.
</div>

> ## Choropleth map with sequential palette

Reviewer: Sorry I changed my mind, the titles look too long. Can you change to just "Choropleth map" and "Heat map" for the headings

<!-- AF example followed: https://analysisfunction.civilservice.gov.uk/policy-store/data-visualisation-colours-in-charts/ -->
```{python}
# | eval: true
# | output: false

import requests
import pandas as pd
import geopandas as gpd
import plotly.graph_objects as go
import io

# Fetch indicator data
url = "https://fingertips.phe.org.uk/api/all_data/csv/by_indicator_id?indicator_ids=41203"
df = pd.read_csv(io.BytesIO(requests.get(url, verify=False).content))

# Keep most recent period
df = df[df["Time period"] == "2023/24"]  # This is the date in the analysis function website.
# df = df.sort_values("Time period").drop_duplicates("Area Code", keep="last")

# Numeric breakpoints for quantiles
bins = [0, 2.4, 2.9, 3.5, 4.6, 12.7]

# Labels in the same order
labels = ["Lowest (0.0 to 2.4)", "(2.4 to 2.9)", "(2.9 to 3.5)", "(3.5 to 4.6)", "Highest (4.6 to 12.7)"]
order = ["Suppressed"] + labels

# Assign bin categories
df["quantile_label"] = pd.cut(df["Value"], bins=bins, labels=labels, include_lowest=True)

# Handle suppressed values (e.g., NaN)

> df["quantile_label"] = df["quantile_label"].astype("category")

Reviewer: `pd.cut` already returns a Categorical series, so this line is redundant

df["quantile_label"] = df["quantile_label"].cat.add_categories("Suppressed")
df["quantile_label"] = df["quantile_label"].fillna("Suppressed")

> df["quantile_label"] = df["quantile_label"].cat.reorder_categories(order, ordered=True)

Reviewer: For brevity and shorter lines, could shorten all this to:
```python
df["quantile_label"] = (
    pd.cut(df["Value"], bins=bins, labels=labels, include_lowest=True)
    .cat.add_categories("Suppressed")
    .fillna("Suppressed")
    .cat.reorder_categories(order, ordered=True)
)
```

    "?outFields=*&where=1%3D1&f=geojson"
)
country_boundaries = gpd.read_file(requests.get(url, verify=False).content).to_crs(4326)

# Get Counties and Unitary Authorities
url = "https://services1.arcgis.com/ESMARspQHYMw9BZ9/arcgis/rest/services/Counties_and_Unitary_Authorities_December_2023_Boundaries_UK_BUC/FeatureServer/0/query?outFields=*&where=1%3D1&f=geojson"
gdf = gpd.read_file(requests.get(url, verify=False).content).to_crs(4326)

# Filter to England only
england_boundaries = country_boundaries[country_boundaries["CTRY23NM"] == "England"]
gdf_clipped_strict = gpd.overlay(gdf, england_boundaries, how="intersection")

merged = gdf_clipped_strict.merge(df, left_on="CTYUA23CD", right_on="Area Code", how="left").reset_index(drop=True)

from afcharts.af_colours import get_af_colours
from afcharts.pio_template import pio

# Get the sequential colour palette in reverse order
sequential = get_af_colours("sequential", number_of_colours=5, include_grey=True)[::-1]

# Set default theme
pio.templates.default = "afcharts"

fig = go.Figure()

cat_to_code = {
    "Highest (4.6 to 12.7)": 5,
    "(3.5 to 4.6)": 4,
    "(2.9 to 3.5)": 3,
    "(2.4 to 2.9)": 2,
    "Lowest (0.0 to 2.4)": 1,

> }

Reviewer: You've already defined the labels, so can just generate the codes automatically e.g.
```suggestion
cat_to_code = {label: i + 1 for i, label in enumerate(labels)}
```

z = merged["quantile_label"].map(cat_to_code)

> # Must add a separate trace for the 'Supressed'.

Reviewer: ```suggestion
# Must add a separate trace for the 'Suppressed'.
```

null_mask = ~pd.isna(z)
fig.add_trace(
    go.Choropleth(
        geojson=merged.loc[null_mask].__geo_interface__,
        locations=merged.loc[null_mask].index,
        z=z[null_mask],
        colorscale=sequential[1:],
        zmin=0,
        zmax=5,
        marker_line_width=0.5,
        marker_line_color="white",
        customdata=merged.loc[null_mask, ["quantile_label", "Value"]],
        hovertext=merged.loc[null_mask, "Area Name"],
        hovertemplate=(
            "<b>%{hovertext}</b><br>"
            "Value: %{customdata[1]}<br>"
            "Category: %{customdata[0]}"
            "<extra></extra>"

>         ),

Reviewer: To reduce the amount of code, I reckon let's just disable all the tooltips.

        showscale=False,  # hide continuous colorbar; we'll add a categorical legend
    )
)

> # Supressed values

Reviewer: ```suggestion
# Suppressed values
```

fig.add_trace(
    go.Choropleth(
        geojson=merged.loc[~null_mask].__geo_interface__,
        locations=merged.loc[~null_mask].index,
        z=[0] * (~null_mask).sum(),  # dummy z
        colorscale=[[0, sequential[0]], [1, sequential[0]]],  # null colour
        marker_line_width=0.5,
        marker_line_color="white",
        customdata=merged.loc[~null_mask, ["quantile_label"]],
        hovertext=merged.loc[~null_mask, "Area Name"],
        hovertemplate=(
            "<b>%{hovertext}</b><br>"
            "Value: Suppressed<br>"
            "Category: %{customdata[0]}"
            "<extra></extra>"
        ),
        showscale=False,
    )
)

> # --- Custom legend (circles + labels) ---

Reviewer: The custom legend is cool but it adds a lot of code just to get circular icons. Another way that uses less code is adding a dummy scatter trace. For example, this works quite well for me:

```python
# Add a legend entry for each category using a dummy scatter trace
for lab, col in zip(order[::-1], sequential[::-1]):
    fig.add_trace(
        go.Scatter(
            x=[None], y=[None],
            mode="markers",
            marker=dict(size=10, color=col, symbol="square"),
            name=lab,
        )
    )
```

and updating `fig.update_layout` with:
```python
fig.update_layout(
    showlegend=True,
    legend=dict(
        title="Equal-sized quintiles of geographies",
        x=0.06,
        y=0.9,
        xanchor="left",
        yanchor="top",
        font=dict(size=14),
    ),
    xaxis=dict(visible=False),
    yaxis=dict(visible=False),
    width=width,
    height=height,
    margin=dict(l=0, r=0, t=0, b=0),
)
```

x = 0.2
y = 1
width = 700
height = 375
width_height_ratio = width / height
item_height = 0.035
r = item_height / 2.0
x_radius = r / width_height_ratio

> y_radius = r

Reviewer: `x_radius` and `y_radius` aren't used

for i, (lab, col) in enumerate(zip(order[::-1], sequential[::-1])):
    y_i = y - i * (item_height + 0.012)

    fig.add_shape(
        type="circle",
        xref="paper",
        yref="paper",
        x0=x,
        x1=x + 2 * r / width_height_ratio,
        y0=y_i - 2 * r,
        y1=y_i,
        line=dict(width=1, color="rgba(0,0,0,0.2)"),
        fillcolor=col,
        layer="above",
    )

    fig.add_annotation(
        x=x + 1.5 * r,
        y=y_i - r,
        xref="paper",
        yref="paper",
        text=str(lab),
        showarrow=False,
        xanchor="left",
        yanchor="middle",
        font=dict(size=12, color="black"),  # ↓ legend label font size
        align="left",
    )

# Add outline of England
fig.add_trace(
    go.Choropleth(
        geojson=england_boundaries.__geo_interface__,
        locations=england_boundaries.index,  # these are indices we put into properties
        z=[0],
        colorscale=[[0, "rgba(0,0,0,0)"], [1, "rgba(0,0,0,0)"]],  # transparent fill
        marker_line_color="black",
        marker_line_width=1,
        hoverinfo="skip",
        showscale=False,
    )
)

> fig.update_layout(showlegend=False, width=width, height=height, margin=dict(l=0, r=0, t=0, b=0))

Reviewer: There's quite a lot of whitespace either side of the map. Is there any easy way to get rid of it? If not, don't worry

Author: Sorry I'm not sure I can do much about it due to the legend (also did not spend much time on it)

Reviewer: Okay no worries

    visible=False,
)
fig.show()
```

<div class="chart-title">Choropleth map with five series of sequential data</div>
<div class="chart-subtitle">Rate of new certifications of visual impairment due to diabetic eye disease in persons aged 12 years and over, per 100,000 population</div>
<div class="chart" aria-describedby="choropleth-desc">

```{python}
#| echo: false

fig.show()

```

</div>
<div class="chart-source">Source: Data used to create this map is available to download from the Fingertips Local Authority health profile produced by the Office for Health Improvement and Disparities.</div>

<div class="chart-alt" id="choropleth-desc">
In England the rate of new certifications of visual impairment (CVI) due to diabetic eye disease in persons aged 12 years and over, were 3.0 per 100,000 population.

The map shows the rate of new certifications of visual impairment (CVI) due to diabetic eye disease in persons aged 12 years and over by upper-tier local authority. The data has been broken down into five equal groups with the highest fifth of areas being in the category showing as darkest blue on the map, and areas with the lowest levels shown in the lightest blue.

> 
> Oldham local authority has the highest rate where 12.7 per 100,000 persons aged 12 years and over were newly certified with a visual impairment due to diabetic eye disease. In contrast City of London local authority has the lowest rate of 0.0 per 100,000 persons aged 12 years and over. 29 local authority areas had values that were suppressed due to a small count.
> </div>

Reviewer: We don't need this long description. Just keep the alt text part

```suggestion
<div class="chart-alt" id="choropleth-desc">
The map shows the rate of new certifications of visual impairment (CVI) due to diabetic eye disease in persons aged 12 years and over by upper-tier local authority. The data has been broken down into five equal groups with the highest fifth of areas being in the category showing as darkest blue on the map, and areas with the lowest levels shown in the lightest blue.
</div>
```

## Heat map with sequential palette
<!-- Source: https://www.gov.uk/government/statistical-data-sets/monthly-domestic-energy-price-statistics -->
```{python}
# | eval: true
# | output: false

import numpy as np
import pandas as pd
import plotly.graph_objects as go

import textwrap

from afcharts.af_colours import get_af_colours
from afcharts.pio_template import pio

# Get the sequential colour palette
sequential = get_af_colours("sequential", number_of_colours=5, include_grey=True)[::-1]

# Set default theme
pio.templates.default = "afcharts"

DATA_URL = "https://assets.publishing.service.gov.uk/media/69f1caf7c42061e837e3abfb/table_211_213__8_.xlsx"

> df = pd.read_excel(DATA_URL, sheet_name="2.1.1", header=11, skiprows=[12])

Reviewer: Can select just the columns we want with `usecols` e.g.
```suggestion
df = pd.read_excel(DATA_URL, sheet_name="2.1.1", header=11, skiprows=[12], usecols="A:B,J:P")
```

Author: speaking of reducing code LOL

df["Year and Period"] = df[["Year and dataset code row", "Quarter"]].astype(str).agg(", ".join, axis=1)

> df = df.set_index(["Year and Period"])

Reviewer: The heatmap is a bit confusing with so many dates in it, can we filter it down?

I tried filtering to just Q1 (Jan to Mar) since 2000 and that looked reasonable.

# Select numeric block

> df = df.iloc[:, 9:-2].copy()

Reviewer: If we only load the columns we need (see `usecols` comment), this can become:

```suggestion
df.drop(columns=["Year and dataset code row", "Quarter"], inplace=True)
```

for c in df.columns:
    df[c] = pd.to_numeric(df[c], errors="coerce")

> 

Reviewer: drop extra blank line
```suggestion
```

null_mask = np.where(df.isna(), 1, np.nan)

fig = go.Figure()

# Wrap each x‑axis label to max 30 characters per line

> wrapped_x = ["<br>".join(textwrap.wrap(col.replace("Real terms price indices:", ""), width=30)) for col in df.columns]

Reviewer: Also remove the notes (e.g. "[Note 3]") and, for All Items, "CPI 2025=100"

> fig.add_trace(go.Heatmap(z=df, x=wrapped_x, y=list(df.index), colorscale=sequential[1:]))

Reviewer: Remove tooltips. The automatic ones are confusing and there's no need to add more code to format them manually

> fig.add_trace(go.Heatmap(z=df, x=wrapped_x, y=list(df.index), colorscale=sequential[1:]))

Reviewer: Vertical/diagonal labels go against AF guidance. Could we flip the heatmap so the labels are on the vertical axis (with horizontal text) with years on the horizontal axis?

fig.update_layout(
    height=600,
    width=600,
    xaxis=dict(
        tickangle=45,
        automargin=True,
    ),
)

# Overlay heatmap for nulls
fig.add_trace(
    go.Heatmap(
        z=null_mask,
        x=wrapped_x,
        y=list(df.index),
        colorscale=[[0, sequential[0]], [1, sequential[0]]],  # null colour
        showscale=False,
        hoverinfo="skip",
    )

> )

Reviewer: I think we can remove this code for null values. I think the latest quarter is the only one with nulls.

We can either just drop this code (I tried and the heatmap looked the same), or we can filter out rows with nulls.

fig.show()
```

<div class="chart-title">Domestic energy price indices in real terms</div>
<div class="chart-subtitle">Heat map of Quarterly consumer price indices of fuel components.</div>
<div class="chart" aria-describedby="heatmap-desc">

```{python}
#| echo: false

fig.show()

```

</div>
<div class="chart-source">Source: https://www.gov.uk/government/statistical-data-sets/monthly-domestic-energy-price-statistics</div>

<div class="chart-alt" id="heatmap-desc">
Figures include VAT where applicable.  VAT rates for coal and coke, gas, electricity and heating oils was 8% from the 2nd quarter of 1994 and 5% since the 4th quarter of 1997 (the rate changed on 1st September). 
Real price series are deflated using GDP (market prices) deflator. The GDP deflator is updated when published by ONS. 
Data may change regularly due to rebasing. The ONS use the latest full year of data as the base year and the Department rebases to 2010=100

Note 1. Aggregate of individual solid fuels, gas, electricity and liquid fuels indices.
Note 2. ULSP, ULSD & motor oil. Other fuels are as defined by ONS. See the Methodology tab for further details.

> Note 3. ONS update the GDP Deflator on a regular basis. As such the real terms price indices are subject to regular revisions to the whole series.

Reviewer: The alt text should describe the chart. The notes and caveats are overkill for an example plot imho. I reckon just direct people to the original source for caveats.
